### ISSUE TITLE
DockerOperator tries to push bytes to XCom, fails to serialize

### ISSUE THREAD

**Issue by MatrixManAtYrService** [apache/airflow#15952]
Apache Airflow version

6f8c204

Environment

OS (e.g. from /etc/os-release): Mac OS 11.3
Kernel: Darwin Kernel Version 20.4.0
Install tools: pip install -e .

### The DAG

```python3
@dag
def hello_docker():
    DockerOperator(
        task_id="say_hi",
        image="bash:latest",
        command=[
            "-c",
            "echo Hello World",
        ],
    )
```

### What Happened
Operator tried to push to XCom, but throwse serialize error
```
[2021-05-19 09:13:40,044] {taskinstance.py:1280} INFO - Exporting the following env vars:
AIRFLOW_CTX_DAG_OWNER=airflow
AIRFLOW_CTX_DAG_ID=docker_mount
AIRFLOW_CTX_TASK_ID=add_one
AIRFLOW_CTX_EXECUTION_DATE=2021-05-19T00:00:00+00:00
AIRFLOW_CTX_DAG_RUN_ID=backfill__2021-05-19T00:00:00+00:00
[2021-05-19 09:13:40,904] {xcom.py:228} ERROR - Could not serialize the XCom value into JSON. If you are using pickle instead of JSON for XCom, then you need to enable pickle support for XCom in your airflow config.
[2021-05-19 09:13:40,904] {taskinstance.py:1481} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/Users/matt/src/airflow/airflow/models/taskinstance.py", line 1137, in _run_raw_task
    self._prepare_and_execute_task_with_callbacks(context, task)
  File "/Users/matt/src/airflow/airflow/models/taskinstance.py", line 1311, in _prepare_and_execute_task_with_callbacks
    result = self._execute_task(context, task_copy)
  File "/Users/matt/src/airflow/airflow/models/taskinstance.py", line 1344, in _execute_task
    self.xcom_push(key=XCOM_RETURN_KEY, value=result)
  File "/Users/matt/src/airflow/airflow/utils/session.py", line 70, in wrapper
    return func(*args, session=session, **kwargs)
  File "/Users/matt/src/airflow/airflow/models/taskinstance.py", line 1919, in xcom_push
    XCom.set(
  File "/Users/matt/src/airflow/airflow/utils/session.py", line 67, in wrapper
    return func(*args, **kwargs)
  File "/Users/matt/src/airflow/airflow/models/xcom.py", line 79, in set
    value = XCom.serialize_value(value)
  File "/Users/matt/src/airflow/airflow/models/xcom.py", line 226, in serialize_value
    return json.dumps(value).encode('UTF-8')
  File "/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type bytes is not JSON serializable
[2021-05-19 09:13:40,906] {taskinstance.py:1524} INFO - Marking task as FAILED. dag_id=docker_mount, task_id=add_one, execution_date=20210519T000000, start_date=20210519T145759, end_date=20210519T151340
[2021-05-19 09:13:40,913] {debug_executor.py:87} ERROR - Failed to execute task: Object of type bytes is not JSON serializable.
```

### What I Expected to Happen

Something useful ended up in XCom, and no errors

**Comment by MatrixManAtYrService** [comment]
@ashb FYI

**Comment by ashb** [comment]
What version of docker provider was this? Did you run against an editable install?

**Comment by MatrixManAtYrService** [comment]
Uh, maybe I should have thought about that beforehand.  I just cloned airflow and did a `pip install -e /path/to/airflow` (so yes, editable).  Let me try again...

I expect these commands to give me a blank slate at a more deterministic version:
```
python3 -m venv venv
source venv/bin/activate
pip install --upgrade pip wheel
pip install apache-airflow apache-airflow-providers-docker
pip freeze | grep apache-airflow
```

At the end, they say this:
```
apache-airflow==2.0.2
apache-airflow-providers-docker==1.2.0
```

Given those versions, I still get the same error.

**Comment by EtsuNDmA** [comment]
https://github.com/apache/airflow/blob/bc004151ed6924ee7bec5d9d047aedb4873806da/airflow/providers/docker/operators/docker.py#L316
I think we should cast self.cli.logs() to string, rather then encoding the line:
```python
ret = self.cli.logs(container=self.container['Id']).decode('utf-8') if self.xcom_all else line
```

**Comment by potiuk** [comment]
I think this is somewhat ambiguous whether the APIs of docker return bytes or strings. Apparently it has been a problem before as  if you look few lines up, the conditional decode happens (and that's why we have line.encode("utf-8") that should stay there.

```
            for line in lines:
                line = line.strip()
                if hasattr(line, 'decode'):
                    # Note that lines returned can also be byte sequences so we have to handle decode here
                    line = line.decode('utf-8')
                res_lines.append(line)
```
Not nice.

I looked at the APIs https://docker-py.readthedocs.io/en/1.2.3/api/  and it is mentioned in both `attach` and `logs` that it will return either `str` or generator. It does not mention generator of what type (and we are using stream=Yes in attach so I guess we get the generator not str). So well, it's kinda not-specified whether the generator returns  bytes or strings and it's "OKeyish" to try to react to both cases.

Interestingly, the documentation says that `attach` method is really a wrapper around logs() - and when you look closer at the method seems what we are trying to do is to retrieve the logs that we already retrieved (and stored in `res_lines` as array of `str`).  So it is likely that we actually receive again the same generator when we call logs after earlier calling "attach" with "stream=True".

So I think the proper fix is actually: 

```
ret = res_lines if self.xcom_all else line
```

However I wonder maybe there was a good reason why the line was encoded to bytes before pushing to Xcom? I think if we fix it, we should make a major release of docker operator.

**Comment by eladkal** [comment]
@MatrixManAtYrService can you check if the issue still reproducible with latest docker provider?
There were some fixes related to Xcoms

**Comment by MatrixManAtYrService** [comment]
@eladkal It appears to have been fixed:

```
❯ pip freeze | grep apache-airflow

    apache-airflow==2.2.5
    apache-airflow-providers-docker==2.5.2
    apache-airflow-providers-ftp==2.1.2
    apache-airflow-providers-http==2.1.2
    apache-airflow-providers-imap==2.2.3
    apache-airflow-providers-sqlite==2.1.3

❯ airflow dags test hello_docker $(date +%Y-%m-%dT%H:%M:%SZ)

    ...
    [2022-04-04 14:32:04,064] {taskinstance.py:1446} INFO - Exporting the following env vars:
    AIRFLOW_CTX_DAG_OWNER=airflow
    AIRFLOW_CTX_DAG_ID=hello_docker
    AIRFLOW_CTX_TASK_ID=say_hi
    AIRFLOW_CTX_EXECUTION_DATE=2022-04-04T14:31:58+00:00
    AIRFLOW_CTX_DAG_RUN_ID=backfill__2022-04-04T14:31:58+00:00
    [2022-04-04 14:32:07,987] {taskinstance.py:1278} INFO - Marking task as SUCCESS. dag_id=hello_docker, task_id=say_hi, execution_date=20220404T143158, start_date=20220404T203159, end_date=20220404T203207
    [2022-04-04 14:32:08,003] {dagrun.py:545} INFO - Marking run <DagRun hello_docker @ 2022-04-04T14:31:58+00:00: backfill__2022-04-04T14:31:58+00:00, externally triggered: False> successful
    [2022-04-04 14:32:08,003] {dagrun.py:590} INFO - DagRun Finished: dag_id=hello_docker, execution_date=2022-04-04T14:31:58+00:00, run_id=backfill__2022-04-04T14:31:58+00:00, run_start_date=2022-04-04 20:31:59.034772+00:00, run_end_date=2022-04-04 20:32:08.003461+00:00, run_duration=8.968689, state=success, external_trigger=False, run_type=backfill, data_interval_start=2022-04-04T14:31:58+00:00, data_interval_end=2022-04-04T14:31:58+00:00, dag_hash=None
    [2022-04-04 14:32:08,003] {backfill_job.py:386} INFO - [backfill progress] | finished run 1 of 1 | tasks waiting: 0 | succeeded: 1 | running: 0 | failed: 0 | skipped: 0 | deadlocked: 0 | not ready: 0
    [2022-04-04 14:32:08,006] {backfill_job.py:851} INFO - Backfill done. Exiting.

❯ sqlite3 airflow.db 'select value from xcom;'

    "Hello World"
```

**Comment by eladkal** [comment]
Great 😃
